src/com/stuypulse/stuylib/math/SLMath.java

```diff
@@ -145,6 +145,128 @@ public static double circular(double x) {
+    /*** INTEGRAL APPROXIMATIONS ***/
+    /*******************************/
+
+    /**
+     * Given data points, calculates integral using the chosen formula.<br>
+     * 
+     * Arrays x and y must be the same length, because they contain data points.
+     * The x-values must be in sorted order.
+     * 
+     * @param x       the x-values of the data points, where (x[i], y[i]) is a point
+     * @param y       the y-values of the data points, where (x[i], y[i]) is a point
+     * @param formula the chosen formula to approximate the integral.
+     * @return approximation of the integral from given points
+     */
+    public static double integrate(double[] x, double[] y, Integral formula) {
+        // Precondition: there always exists x[i], y[i]
+        if(x.length != y.length) {
+            throw new IllegalArgumentException("x and y-value arrays should be the same length");
+        }
+
+        // Precondition: x-values should be sorted
+        for(int i = 1; i < x.length; ++i) {
+            if(x[i] < x[i - 1]) {
+                throw new IllegalArgumentException("x-values should be in sorted order");
+            }
+        }
+
+        return formula.calculate(x, y);
+    }
+
+    public static final Integral DEFAULT_INTEGRAL = new Integral.Simpson();
+    /**
+     * Calls integrate(double[] x, double[] y, Integral formula) with default formula Simpson.
+     * 
+     * @param x the x-values of the data points, where (x[i], y[i]) is a point
+     * @param y the y-values of the data points, where (x[i], y[i]) is a point
+     * @return approximation of th eintegral from the given points using the Simpson formula.
+     */
+    public static double integrate(double[] x, double[] y) {
+        return integrate(x, y, DEFAULT_INTEGRAL);
+    }
+
+    /**
+     * Integral is an interface used by integrate() to get the method used for numerical integration.  
+     * It also contains several classes containing various formulas for integration.
+     * 
+     * Riemann: Uses the Riemann sum to approximate the integral. (https://en.wikipedia.org/wiki/Riemann_sum)<br>
+     * Trapezoidal: Uses the trapezoidal rule to approximate the integral. (https://en.wikipedia.org/wiki/Trapezoidal_rule)<br>
+     * Simpson: Uses Composite Simpson's Rule for irregularly spaced data to approximate the integral. (https://en.wikipedia.org/wiki/Simpson%27s_rule)<br>
+     */
+    public interface Integral {
+        /**
+         * Given data points, gives an approximation for the integral.
+         * Arrays x and y must be the same length, because they contain data points.
+         * The x-values must be in sorted order.
+         * 
+         * @param x the x-values of the data points, where (x[i], y[i]) is a point
+         * @param y the y-values of the data points, where (x[i], y[i]) is a point
+         * @return approximation of the integral from given points
+         */
+        public double calculate(double[] x, double[] y);
```

Comment:
Vector2D instead of X, Y. Also it might be better if the integral keeps an internal list, then you can call a .update(Vector2D) to give it a new value, I feel like it would make more sense for an integral class. It would remove the need for the array

however, you could still use an array through the use of a static helper function that could call update on each item of the array

This would makes its use in PID easier because you wouldnt need an integral double, boolean for checking if it has the past error, and it would just be a lot easier.

this would also make using the simpson method of integrals possible in PID